x/stakeibc/keeper/update_redemption_rates_test.go

```diff
@@ -0,0 +1,272 @@
+	stakeibc "github.com/Stride-Labs/stride/x/stakeibc/types"
+)
+
+type UpdateRedemptionRatesTestCase struct {
+	host_zone          stakeibc.HostZone
+	redemption_rate_t0 sdk.Dec
+	staked_bal         uint64
+	undelegated_bal    uint64
+	just_deposited_bal uint64
+	st_supply          uint64
+}
+
+func (s *KeeperTestSuite) SetupUpdateRedemptionRates(
+	STAKED_BAL uint64,
+	UNDELEGATED_BAL uint64,
+	JUST_DEPOSITED_BAL uint64,
+	ST_SUPPLY uint64,
+	REDEMPTION_RATE_T0 sdk.Dec,
+) UpdateRedemptionRatesTestCase {
+
+	// add some deposit records with status STAKE
+	//    to comprise the undelegated delegation account balance i.e. "to be staked"
+	toBeStakedDepositRecord := recordtypes.DepositRecord{
+		HostZoneId: "GAIA",
+		Amount:     int64(UNDELEGATED_BAL),
+		Status:     recordtypes.DepositRecord_STAKE,
+	}
+	s.App.RecordsKeeper.AppendDepositRecord(s.Ctx, toBeStakedDepositRecord)
+
+	// add a balance to the stakeibc module account (via records)
+	//    to comprise the stakeibc module account balance i.e. "to be transferred"
+	toBeTransferedDepositRecord := recordtypes.DepositRecord{
+		HostZoneId: "GAIA",
+		Amount:     int64(JUST_DEPOSITED_BAL),
+		Status:     recordtypes.DepositRecord_TRANSFER,
+	}
+	s.App.RecordsKeeper.AppendDepositRecord(s.Ctx, toBeTransferedDepositRecord)
+
+	// set the stSupply by minting to a random user account
+	user := Account{
+		acc:           s.TestAccs[0],
+		stAtomBalance: sdk.NewInt64Coin(stAtom, int64(ST_SUPPLY)),
+	}
+	s.FundAccount(user.acc, user.stAtomBalance)
+
+	// set the staked balance on the host zone
+	hostZone := stakeibc.HostZone{
+		ChainId:        "GAIA",
+		HostDenom:      "uatom",
+		StakedBal:      STAKED_BAL,
+		RedemptionRate: REDEMPTION_RATE_T0,
+	}
+	s.App.StakeibcKeeper.SetHostZone(s.Ctx, hostZone)
+
+	return UpdateRedemptionRatesTestCase{
+		host_zone:          hostZone,
+		redemption_rate_t0: REDEMPTION_RATE_T0,
+		staked_bal:         STAKED_BAL,
+		undelegated_bal:    UNDELEGATED_BAL,
+		just_deposited_bal: JUST_DEPOSITED_BAL,
+		st_supply:          ST_SUPPLY,
+	}
+}
+
+func (s *KeeperTestSuite) TestUpdateRedemptionRatesSuccessful() {
+
+	STAKED_BAL := uint64(5)
+	UNDELEGATED_BAL := uint64(3)
+	JUST_DEPOSITED_BAL := uint64(3)
+	ST_SUPPLY := uint64(10)
+
+	REDEMPTION_RATE_T0 := sdk.NewDec(1)
+	tc := s.SetupUpdateRedemptionRates(STAKED_BAL, UNDELEGATED_BAL, JUST_DEPOSITED_BAL, ST_SUPPLY, REDEMPTION_RATE_T0)
+
+	// sanity check on inputs (check redemptionRate at genesis is 1)
+	s.Require().Equal(tc.redemption_rate_t0, sdk.NewDec(1))
+
+	records := s.App.RecordsKeeper.GetAllDepositRecord(s.Ctx)
+	s.App.StakeibcKeeper.UpdateRedemptionRates(s.Ctx, records)
+
+	hz, found := s.App.StakeibcKeeper.GetHostZone(s.Ctx, tc.host_zone.ChainId)
+	s.Require().True(found)
+	rrNew := hz.RedemptionRate
+
+	s.Require().Equal(rrNew, sdk.NewDec(int64(STAKED_BAL)+int64(UNDELEGATED_BAL)+int64(JUST_DEPOSITED_BAL)).Quo(sdk.NewDec(int64(ST_SUPPLY))))
+	s.Require().Equal(rrNew, sdk.NewDec(11).Quo(sdk.NewDec(10)))
```

Comment:
i'd be careful about including a copy of the code that's being tested, in the test itself (i.e. this equation). I see you have the hard coded case below though!

Also do you have to do `sdk.NewDec(11).Quo(sdk.NewDec(10))` because `NewDec` takes an int as a parameter?

Reply:
right, NewDec only takes ints so afaiu the best way to get 1.1 is using 11 Quo 10

Any ideas on how to test randomized input tests without "including a copy of the code that's being tested, in the test itself (i.e. this equation)" ? @sampocs

Reply:
I think it's fine to use the equation for that! Especially since you will already have plenty of coverage with hard coded numbers from the other tests